File: html-generators/generate.py
#!/usr/bin/env python3
"""Generate HTML detail pages from JSON snippet files and slug-template.html."""

import json
import glob
import os
import html
import re
from urllib.parse import quote

BASE_URL = "https://javaevolved.github.io"
TEMPLATE_FILE = "templates/slug-template.html"
WHY_CARD_TEMPLATE = "templates/why-card.html"
RELATED_CARD_TEMPLATE = "templates/related-card.html"
SOCIAL_SHARE_TEMPLATE = "templates/social-share.html"
DOC_LINK_TEMPLATE = "templates/doc-link.html"
INDEX_TEMPLATE = "templates/index.html"
INDEX_CARD_TEMPLATE = "templates/index-card.html"
CONTENT_DIR = "content"
SITE_DIR = "site"

CATEGORY_DISPLAY = {
    "language": "Language",
    "collections": "Collections",
    "strings": "Strings",
    "streams": "Streams",
    "concurrency": "Concurrency",
    "io": "I/O",
    "errors": "Errors",
    "datetime": "Date/Time",
    "security": "Security",
    "tooling": "Tooling",
    "enterprise": "Enterprise",
}


def escape(text):
    """HTML-escape text for use in attributes and content."""
    return html.escape(text, quote=True)


def replace_tokens(template, replacements):
    """Replace {{token}} placeholders in a template string."""
    def replacer(m):
        key = m.group(1)
        return replacements.get(key, m.group(0))
    return re.sub(r"\{\{(\w+)\}\}", replacer, template)


def json_escape(text):
    """Escape text for embedding in JSON strings inside ld+json blocks.

    Uses ASCII-only encoding to match the original HTML files which use
    \\uXXXX escapes for non-ASCII characters in ld+json blocks.
    """
    return json.dumps(text, ensure_ascii=True)[1:-1]


def load_template():
    """Load the external HTML template."""
    with open(TEMPLATE_FILE) as f:
        return f.read()


def load_all_snippets():
    """Load all JSON snippet files, keyed by category/slug."""
    snippets = {}
    categories = [
        "language", "collections", "strings", "streams", "concurrency",
        "io", "errors", "datetime", "security", "tooling", "enterprise",
    ]
    json_files = []
    for cat in categories:
        json_files.extend(sorted(glob.glob(f"{CONTENT_DIR}/{cat}/*.json")))
    for path in json_files:
        with open(path) as f:
            data = json.load(f)
        key = f"{data['category']}/{data['slug']}"
        data["_path"] = key
        snippets[key] = data
    return snippets


def render_nav_arrows(data):
    """Render prev/next navigation arrows."""
    parts = []
    if data.get("prev"):
        parts.append(
            f'<a href="/{data["prev"]}.html" aria-label="Previous pattern">←</a>'
        )
    else:
        parts.append('<span class="nav-arrow-disabled">←</span>')
    if data.get("next"):
        parts.append(
            f'<a href="/{data["next"]}.html" aria-label="Next pattern">→</a>'
        )
    else:
        parts.append("")
    return "\n          ".join(parts)


def render_why_cards(why_card_template, why_list):
    """Render the 3 why-modern-wins cards."""
    cards = []
    for w in why_list:
        cards.append(
            replace_tokens(why_card_template, {
                "icon": w["icon"],
                "title": escape(w["title"]),
                "desc": escape(w["desc"]),
            })
        )
    return "\n".join(cards)


def render_related_card(related_card_template, related_data):
    """Render a single related pattern tip-card."""
    cat = related_data["category"]
    cat_display = CATEGORY_DISPLAY[cat]

    return replace_tokens(related_card_template, {
        "category": cat,
        "slug": related_data["slug"],
        "catDisplay": cat_display,
        "difficulty": related_data["difficulty"],
        "title": escape(related_data["title"]),
        "oldLabel": escape(related_data["oldLabel"]),
        "oldCode": escape(related_data["oldCode"]),
        "modernLabel": escape(related_data["modernLabel"]),
        "modernCode": escape(related_data["modernCode"]),
        "jdkVersion": related_data["jdkVersion"],
    })


def render_related_section(related_card_template, related_paths, all_snippets):
    """Render all related pattern cards."""
    cards = []
    for path in related_paths:
        if path in all_snippets:
            cards.append(render_related_card(related_card_template, all_snippets[path]))
    return "\n".join(cards)


def render_social_share(social_share_template, slug, title):
    """Render social share URLs using the old flat URL format."""
    page_url = f"{BASE_URL}/{slug}.html"
    share_text = f"{title} \u2013 java.evolved"

    encoded_url = quote(page_url, safe="")
    encoded_text = quote(share_text, safe="")

    return replace_tokens(social_share_template, {
        "encodedUrl": encoded_url,
        "encodedText": encoded_text,
    })


def render_doc_links(doc_link_template, docs):
    """Render documentation links."""
    return "\n".join(
        replace_tokens(doc_link_template, {
            "docTitle": escape(d["title"]),
            "docHref": d["href"],
        })
        for d in docs
    )


def _support_badge(state):
    return {"preview": "Preview", "experimental": "Experimental"}.get(state, "Available")


def _support_badge_class(state):
    return {"preview": "preview", "experimental": "experimental"}.get(state, "widely")


def render_index_card(index_card_template, data):
    """Render a single index page preview card."""
    cat = data["category"]
    return replace_tokens(index_card_template, {
        "category": cat,
        "slug": data["slug"],
        "catDisplay": CATEGORY_DISPLAY[cat],
        "title": escape(data["title"]),
        "oldCode": escape(data["oldCode"]),
        "modernCode": escape(data["modernCode"]),
        "jdkVersion": data["jdkVersion"],
    })


def generate_html(template, why_card_template, related_card_template,
                  social_share_template, doc_link_template, data, all_snippets):
    """Generate the full HTML page for a snippet by rendering the template."""
    cat = data["category"]
    slug = data["slug"]
    cat_display = CATEGORY_DISPLAY[cat]

    # Build the substitution map
    replacements = {
        "title": escape(data["title"]),
        "summary": escape(data["summary"]),
        "slug": slug,
        "category": cat,
        "categoryDisplay": cat_display,
        "difficulty": data["difficulty"],
        "jdkVersion": data["jdkVersion"],
        "oldLabel": escape(data["oldLabel"]),
        "modernLabel": escape(data["modernLabel"]),
        "oldCode": escape(data["oldCode"]),
        "modernCode": escape(data["modernCode"]),
        "oldApproach": escape(data["oldApproach"]),
        "modernApproach": escape(data["modernApproach"]),
        "explanation": escape(data["explanation"]),
        "supportDescription": escape(data["support"]["description"]),
        "supportBadge": _support_badge(data["support"]["state"]),
        "supportBadgeClass": _support_badge_class(data["support"]["state"]),
        "canonicalUrl": f"{BASE_URL}/{cat}/{slug}.html",
        "flatUrl": f"{BASE_URL}/{slug}.html",
        "titleJson": json_escape(data["title"]),
        "summaryJson": json_escape(data["summary"]),
        "categoryDisplayJson": json_escape(cat_display),
        "navArrows": render_nav_arrows(data),
        "whyCards": render_why_cards(why_card_template, data["whyModernWins"]),
        "docLinks": render_doc_links(doc_link_template, data.get("docs", [])),
        "relatedCards": render_related_section(related_card_template, data.get("related", []), all_snippets),
        "socialShare": render_social_share(social_share_template, slug, data["title"]),
    }

    return replace_tokens(template, replacements)


def main():
    template = load_template()
    why_card_template = open(WHY_CARD_TEMPLATE).read()
    related_card_template = open(RELATED_CARD_TEMPLATE).read()
    social_share_template = open(SOCIAL_SHARE_TEMPLATE).read()
    doc_link_template = open(DOC_LINK_TEMPLATE).read()
    index_template = open(INDEX_TEMPLATE).read()
    index_card_template = open(INDEX_CARD_TEMPLATE).read()
    all_snippets = load_all_snippets()
    print(f"Loaded {len(all_snippets)} snippets")

    for key, data in all_snippets.items():
        html_content = generate_html(
            template, why_card_template, related_card_template,
            social_share_template, doc_link_template, data, all_snippets
        ).strip()
        out_dir = os.path.join(SITE_DIR, data['category'])
        os.makedirs(out_dir, exist_ok=True)
        out_path = os.path.join(out_dir, f"{data['slug']}.html")
        with open(out_path, "w", newline="") as f:
            f.write(html_content)

    print(f"Generated {len(all_snippets)} HTML files")

    # Rebuild data/snippets.json from individual JSON files
    # This file is used at runtime by app.js for search
    snippets_list = []
    for key, data in all_snippets.items():
        entry = {k: v for k, v in data.items() if k not in ("_path", "prev", "next", "related")}
        snippets_list.append(entry)

    os.makedirs(os.path.join(SITE_DIR, "data"), exist_ok=True)
    with open(os.path.join(SITE_DIR, "data", "snippets.json"), "w") as f:
        json.dump(snippets_list, f, indent=2, ensure_ascii=False)
        f.write("\n")

    print(f"Rebuilt data/snippets.json with {len(snippets_list)} entries")

    # Generate index.html from template
    tip_cards = "\n".join(
        render_index_card(index_card_template, data)
        for data in all_snippets.values()
    )
    count = len(all_snippets)
    index_html = replace_tokens(index_template, {
        "tipCards": tip_cards,
        "snippetCount": str(count),
    })
    with open(os.path.join(SITE_DIR, "index.html"), "w") as f:
        f.write(index_html)

    print(f"Generated index.html with {count} cards")


if __name__ == "__main__":
    main()

File: html-generators/generate.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25
//DEPS com.fasterxml.jackson.core:jackson-databind:2.18.3

import module java.base;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.*;

/**
 * Generate HTML detail pages from JSON snippet files and slug-template.html.
 * JBang equivalent of generate.py — produces identical output.
 */
static final String BASE_URL = "https://javaevolved.github.io";
static final String CONTENT_DIR = "content";
static final String SITE_DIR = "site";
static final Pattern TOKEN = Pattern.compile("\\{\\{(\\w+)}}");
static final ObjectMapper MAPPER = new ObjectMapper();

static final String CATEGORIES_FILE = "html-generators/categories.properties";
static final SequencedMap<String, String> CATEGORY_DISPLAY = loadCategoryDisplay();

static SequencedMap<String, String> loadCategoryDisplay() {
    try {
        var map = new LinkedHashMap<String, String>();
        for (var line : Files.readAllLines(Path.of(CATEGORIES_FILE))) {
            line = line.strip();
            if (line.isEmpty() || line.startsWith("#")) continue;
            var idx = line.indexOf('=');
            if (idx > 0) map.put(line.substring(0, idx).strip(), line.substring(idx + 1).strip());
        }
        return map;
    } catch (IOException e) {
        throw new UncheckedIOException(e);
    }
}

static final Set<String> EXCLUDED_KEYS = Set.of("_path", "prev", "next", "related");

record Snippet(JsonNode node) {
    String get(String f)    { return node.get(f).asText(); }
    String slug()           { return get("slug"); }
    String category()       { return get("category"); }
    String title()          { return get("title"); }
    String summary()        { return get("summary"); }
    String difficulty()     { return get("difficulty"); }
    String jdkVersion()     { return get("jdkVersion"); }
    String oldLabel()       { return get("oldLabel"); }
    String modernLabel()    { return get("modernLabel"); }
    String oldCode()        { return get("oldCode"); }
    String modernCode()     { return get("modernCode"); }
    String oldApproach()    { return get("oldApproach"); }
    String modernApproach() { return get("modernApproach"); }
    String explanation()    { return get("explanation"); }
    String supportState()   { return node.get("support").get("state").asText(); }
    String supportDesc()    { return node.get("support").get("description").asText(); }
    String key()            { return category() + "/" + slug(); }
    String catDisplay()     { return CATEGORY_DISPLAY.get(category()); }
    JsonNode whyModernWins() { return node.get("whyModernWins"); }

    Optional<String> optText(String f) {
        var n = node.get(f);
        return n != null && !n.isNull() ? Optional.of(n.asText()) : Optional.empty();
    }

    List<String> related() {
        var rel = node.get("related");
        if (rel == null) return List.of();
        var paths = new ArrayList<String>();
        rel.forEach(n -> paths.add(n.asText()));
        return paths;
    }
}

record Templates(String page, String whyCard, String relatedCard, String socialShare,
                 String index, String indexCard, String docLink) {
    static Templates load() throws IOException {
        return new Templates(
            Files.readString(Path.of("templates/slug-template.html")),
            Files.readString(Path.of("templates/why-card.html")),
            Files.readString(Path.of("templates/related-card.html")),
            Files.readString(Path.of("templates/social-share.html")),
            Files.readString(Path.of("templates/index.html")),
            Files.readString(Path.of("templates/index-card.html")),
            Files.readString(Path.of("templates/doc-link.html")));
    }
}

void main() throws IOException {
    var templates = Templates.load();
    var allSnippets = loadAllSnippets();
    IO.println("Loaded %d snippets".formatted(allSnippets.size()));

    for (var snippet : allSnippets.values()) {
        var html = generateHtml(templates, snippet, allSnippets).strip();
        Files.createDirectories(Path.of(SITE_DIR, snippet.category()));
        Files.writeString(Path.of(SITE_DIR, snippet.category(), snippet.slug() + ".html"), html);
    }
    IO.println("Generated %d HTML files".formatted(allSnippets.size()));

    // Rebuild data/snippets.json
    var snippetsList = allSnippets.values().stream()
            .map(s -> {
                Map<String, Object> map = MAPPER.convertValue(s.node(), new TypeReference<LinkedHashMap<String, Object>>() {});
                EXCLUDED_KEYS.forEach(map::remove);
                return map;
            })
            .toList();

    Files.createDirectories(Path.of(SITE_DIR, "data"));
    var prettyMapper = new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT);
    Files.writeString(Path.of(SITE_DIR, "data", "snippets.json"), prettyMapper.writeValueAsString(snippetsList) + "\n");
    IO.println("Rebuilt data/snippets.json with %d entries".formatted(snippetsList.size()));

    // Generate index.html from template
    var tipCards = allSnippets.values().stream()
            .map(s -> renderIndexCard(templates.indexCard(), s))
            .collect(Collectors.joining("\n"));
    var indexHtml = replaceTokens(templates.index(), Map.of(
            "tipCards", tipCards,
            "snippetCount", String.valueOf(allSnippets.size())));
    Files.writeString(Path.of(SITE_DIR, "index.html"), indexHtml);
    IO.println("Generated index.html with %d cards".formatted(allSnippets.size()));
}

SequencedMap<String, Snippet> loadAllSnippets() throws IOException {
    SequencedMap<String, Snippet> snippets = new LinkedHashMap<>();
    for (var cat : CATEGORY_DISPLAY.sequencedKeySet()) {
        var catDir = Path.of(CONTENT_DIR, cat);
        if (!Files.isDirectory(catDir)) continue;
        try (var stream = Files.newDirectoryStream(catDir, "*.json")) {
            var sorted = new ArrayList<Path>();
            stream.forEach(sorted::add);
            sorted.sort(Path::compareTo);
            for (var path : sorted) {
                var snippet = new Snippet(MAPPER.readTree(path.toFile()));
                snippets.put(snippet.key(), snippet);
            }
        }
    }
    return snippets;
}

String escape(String text) {
    return text == null ? "" : text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;").replace("'", "&#x27;");
}

String jsonEscape(String text) throws IOException {
    var quoted = MAPPER.writeValueAsString(text);
    var inner = quoted.substring(1, quoted.length() - 1);
    var sb = new StringBuilder(inner.length());
    for (int i = 0; i < inner.length(); i++) {
        char c = inner.charAt(i);
        sb.append(c > 127 ? "\\u%04x".formatted((int) c) : String.valueOf(c));
    }
    return sb.toString();
}

String urlEncode(String s) {
    return URLEncoder.encode(s, StandardCharsets.UTF_8).replace("+", "%20");
}

String supportBadge(String state) {
    return switch (state) {
        case "preview" -> "Preview";
        case "experimental" -> "Experimental";
        default -> "Available";
    };
}

String supportBadgeClass(String state) {
    return switch (state) {
        case "preview" -> "preview";
        case "experimental" -> "experimental";
        default -> "widely";
    };
}

String renderNavArrows(Snippet snippet) {
    var prev = snippet.optText("prev")
            .map(p -> "<a href=\"/%s.html\" aria-label=\"Previous pattern\">←</a>".formatted(p))
            .orElse("<span class=\"nav-arrow-disabled\">←</span>");
    var next = snippet.optText("next")
            .map(n -> "<a href=\"/%s.html\" aria-label=\"Next pattern\">→</a>".formatted(n))
            .orElse("");
    return prev + "\n          " + next;
}

String renderIndexCard(String tpl, Snippet s) {
    return replaceTokens(tpl, Map.ofEntries(
            Map.entry("category", s.category()), Map.entry("slug", s.slug()),
            Map.entry("catDisplay", s.catDisplay()), Map.entry("title", escape(s.title())),
            Map.entry("oldCode", escape(s.oldCode())), Map.entry("modernCode", escape(s.modernCode())),
            Map.entry("jdkVersion", s.jdkVersion())));
}

String renderWhyCards(String tpl, JsonNode whyList) {
    var cards = new ArrayList<String>();
    for (var w : whyList)
        cards.add(replaceTokens(tpl, Map.of(
                "icon", w.get("icon").asText(),
                "title", escape(w.get("title").asText()),
                "desc", escape(w.get("desc").asText()))));
    return String.join("\n", cards);
}

String renderRelatedCard(String tpl, Snippet rel) {
    return replaceTokens(tpl, Map.ofEntries(
            Map.entry("category", rel.category()), Map.entry("slug", rel.slug()),
            Map.entry("catDisplay", rel.catDisplay()), Map.entry("difficulty", rel.difficulty()),
            Map.entry("title", escape(rel.title())),
            Map.entry("oldLabel", escape(rel.oldLabel())), Map.entry("oldCode", escape(rel.oldCode())),
            Map.entry("modernLabel", escape(rel.modernLabel())), Map.entry("modernCode", escape(rel.modernCode())),
            Map.entry("jdkVersion", rel.jdkVersion())));
}

String renderDocLinks(String tpl, JsonNode docs) {
    var links = new ArrayList<String>();
    for (var d : docs)
        links.add(replaceTokens(tpl, Map.of(
                "docTitle", escape(d.get("title").asText()),
                "docHref", d.get("href").asText())));
    return String.join("\n", links);
}

String renderRelatedSection(String tpl, Snippet snippet, Map<String, Snippet> all) {
    return snippet.related().stream().filter(all::containsKey)
            .map(p -> renderRelatedCard(tpl, all.get(p)))
            .collect(Collectors.joining("\n"));
}

String renderSocialShare(String tpl, String slug, String title) {
    var encodedUrl = urlEncode("%s/%s.html".formatted(BASE_URL, slug));
    var encodedText = urlEncode("%s \u2013 java.evolved".formatted(title));
    return replaceTokens(tpl, Map.of("encodedUrl", encodedUrl, "encodedText", encodedText));
}

String generateHtml(Templates tpl, Snippet s, Map<String, Snippet> all) throws IOException {
    return replaceTokens(tpl.page(), Map.ofEntries(
            Map.entry("title", escape(s.title())), Map.entry("summary", escape(s.summary())),
            Map.entry("slug", s.slug()), Map.entry("category", s.category()),
            Map.entry("categoryDisplay", s.catDisplay()), Map.entry("difficulty", s.difficulty()),
            Map.entry("jdkVersion", s.jdkVersion()),
            Map.entry("oldLabel", escape(s.oldLabel())), Map.entry("modernLabel", escape(s.modernLabel())),
            Map.entry("oldCode", escape(s.oldCode())), Map.entry("modernCode", escape(s.modernCode())),
            Map.entry("oldApproach", escape(s.oldApproach())), Map.entry("modernApproach", escape(s.modernApproach())),
            Map.entry("explanation", escape(s.explanation())),
            Map.entry("supportDescription", escape(s.supportDesc())),
            Map.entry("supportBadge", supportBadge(s.supportState())),
            Map.entry("supportBadgeClass", supportBadgeClass(s.supportState())),
            Map.entry("canonicalUrl", "%s/%s/%s.html".formatted(BASE_URL, s.category(), s.slug())),
            Map.entry("flatUrl", "%s/%s.html".formatted(BASE_URL, s.slug())),
            Map.entry("titleJson", jsonEscape(s.title())), Map.entry("summaryJson", jsonEscape(s.summary())),
            Map.entry("categoryDisplayJson", jsonEscape(s.catDisplay())),
            Map.entry("navArrows", renderNavArrows(s)),
            Map.entry("whyCards", renderWhyCards(tpl.whyCard(), s.whyModernWins())),
            Map.entry("docLinks", renderDocLinks(tpl.docLink(), s.node().withArray("docs"))),
            Map.entry("relatedCards", renderRelatedSection(tpl.relatedCard(), s, all)),
            Map.entry("socialShare", renderSocialShare(tpl.socialShare(), s.slug(), s.title()))));
}

String replaceTokens(String template, Map<String, String> replacements) {
    var m = TOKEN.matcher(template);
    var sb = new StringBuilder();
    while (m.find()) m.appendReplacement(sb, Matcher.quoteReplacement(replacements.getOrDefault(m.group(1), m.group(0))));
    m.appendTail(sb);
    return sb.toString();
}
